File: analyze_gap_scan.py
#!/usr/bin/env python3
"""
DAMSA Gap Scan Results Analysis
Reads gap scan results and creates visualization plots
"""

import numpy as np
import matplotlib.pyplot as plt
from matplotlib.gridspec import GridSpec
import sys
import os

def read_gap_scan_results(filename='gap_scan_results.txt'):
    """Read the gap scan results file"""
    if not os.path.exists(filename):
        print(f"Error: {filename} not found!")
        print("Please run the gap scan simulation first.")
        sys.exit(1)
    
    # Read all lines, skip header lines and comments
    data = []
    with open(filename, 'r') as f:
        for line in f:
            line = line.strip()
            # Skip empty lines and comment lines
            if not line or line.startswith('#'):
                continue
            # Parse data line
            parts = line.split()
            if len(parts) >= 8:
                try:
                    row = [float(x) for x in parts[:8]]
                    data.append(row)
                except ValueError:
                    continue
    
    if not data:
        print("Error: No data found in file!")
        sys.exit(1)
    
    data = np.array(data)
    
    results = {
        'gap': data[:, 0],
        'csi_z': data[:, 1],
        'target_photons': data[:, 2],
        'target_neutrons': data[:, 3],
        'detector_photons': data[:, 4],
        'detector_neutrons': data[:, 5],
        'forward_photons': data[:, 6],
        'efficiency': data[:, 7]
    }
    
    return results

def plot_gap_scan_results(results, output_prefix='gap_scan'):
    """Create comprehensive visualization of gap scan results"""
    
    fig = plt.figure(figsize=(16, 10))
    gs = GridSpec(2, 3, figure=fig, hspace=0.3, wspace=0.3)
    
    optimal_idx = np.argmax(results['efficiency'])
    optimal_gap = results['gap'][optimal_idx]
    optimal_eff = results['efficiency'][optimal_idx]
    
    ax1 = fig.add_subplot(gs[0, 0])
    ax1.plot(results['gap'], results['efficiency'], 'o-', color='blue', 
             linewidth=2, markersize=8, label='Photon Efficiency')
    ax1.set_xlabel('Gap Distance (cm)', fontsize=12)
    ax1.set_ylabel('Efficiency (%)', fontsize=12)
    ax1.set_title('Photon Detection Efficiency vs Gap', fontsize=14, fontweight='bold')
    ax1.grid(True, alpha=0.3)
    ax1.axvline(optimal_gap, color='red', linestyle='--', alpha=0.5, 
                label=f'Optimal: {optimal_gap:.0f} cm')
    ax1.plot(optimal_gap, optimal_eff, 'r*', markersize=20)
    ax1.legend()
    
    ax2 = fig.add_subplot(gs[0, 1])
    ax2.plot(results['gap'], results['detector_photons'], 'o-', color='green', 
             linewidth=2, markersize=8)
    ax2.set_xlabel('Gap Distance (cm)', fontsize=12)
    ax2.set_ylabel('Number of Photons', fontsize=12)
    ax2.set_title('Photons Reaching Detector vs Gap', fontsize=14, fontweight='bold')
    ax2.grid(True, alpha=0.3)
    ax2.axvline(optimal_gap, color='red', linestyle='--', alpha=0.5)
    
    ax3 = fig.add_subplot(gs[0, 2])
    ax3.plot(results['gap'], results['forward_photons'], 'o-', color='purple', 
             linewidth=2, markersize=8)
    ax3.set_xlabel('Gap Distance (cm)', fontsize=12)
    ax3.set_ylabel('Number of Forward Photons', fontsize=12)
    ax3.set_title('Forward Photons (0-20°) vs Gap', fontsize=14, fontweight='bold')
    ax3.grid(True, alpha=0.3)
    ax3.axvline(optimal_gap, color='red', linestyle='--', alpha=0.5)
    
    ax4 = fig.add_subplot(gs[1, 0])
    ax4.plot(results['gap'], results['detector_neutrons'], 'o-', color='orange', 
             linewidth=2, markersize=8)
    ax4.set_xlabel('Gap Distance (cm)', fontsize=12)
    ax4.set_ylabel('Number of Neutrons', fontsize=12)
    ax4.set_title('Neutrons Reaching Detector vs Gap', fontsize=14, fontweight='bold')
    ax4.grid(True, alpha=0.3)
    ax4.axvline(optimal_gap, color='red', linestyle='--', alpha=0.5)
    
    ax5 = fig.add_subplot(gs[1, 1])
    ax5.plot(results['gap'], results['target_photons'], 'o-', color='blue', 
             linewidth=2, markersize=8, label='Target Exit Photons', alpha=0.7)
    ax5.plot(results['gap'], results['detector_photons'], 's-', color='green', 
             linewidth=2, markersize=8, label='Detector Photons', alpha=0.7)
    ax5.set_xlabel('Gap Distance (cm)', fontsize=12)
    ax5.set_ylabel('Number of Photons', fontsize=12)
    ax5.set_title('Target Exit vs Detector Photons', fontsize=14, fontweight='bold')
    ax5.grid(True, alpha=0.3)
    ax5.legend()
    ax5.axvline(optimal_gap, color='red', linestyle='--', alpha=0.5)
    
    ax6 = fig.add_subplot(gs[1, 2])
    ax6.axis('off')
    
    table_data = []
    table_data.append(['Parameter', 'Value'])
    table_data.append(['Optimal Gap', f'{optimal_gap:.0f} cm'])
    table_data.append(['Optimal Efficiency', f'{optimal_eff:.2f}%'])
    table_data.append(['CsI Position', f'{results["csi_z"][optimal_idx]:.1f} cm'])
    table_data.append(['Detector Photons', f'{int(results["detector_photons"][optimal_idx])}'])
    table_data.append(['Forward Photons', f'{int(results["forward_photons"][optimal_idx])}'])
    
    table = ax6.table(cellText=table_data, cellLoc='left', loc='center',
                     colWidths=[0.6, 0.4])
    table.auto_set_font_size(False)
    table.set_fontsize(10)
    table.scale(1, 2)
    
    for i in range(2):
        table[(0, i)].set_facecolor('#4CAF50')
        table[(0, i)].set_text_props(weight='bold', color='white')
    
    ax6.set_title('Optimal Configuration', fontsize=14, fontweight='bold', pad=20)
    
    plt.suptitle('DAMSA Gap Distance Optimization Scan Results', 
                 fontsize=16, fontweight='bold', y=0.98)
    
    plt.savefig(f'{output_prefix}_analysis.png', dpi=300, bbox_inches='tight')
    print(f"Analysis plot saved as: {output_prefix}_analysis.png")
    
    return optimal_gap, optimal_eff

def print_recommendations(results):
    """Print detailed recommendations based on the scan"""
    
    optimal_idx = np.argmax(results['efficiency'])
    optimal_gap = results['gap'][optimal_idx]
    
    print("\n" + "="*60)
    print("DAMSA GAP SCAN RECOMMENDATIONS")
    print("="*60)
    print(f"\nOptimal Gap Distance: {optimal_gap:.0f} cm")
    print(f"Corresponding CsI Center Z: {results['csi_z'][optimal_idx]:.1f} cm")
    print(f"Maximum Efficiency: {results['efficiency'][optimal_idx]:.2f}%")
    print(f"Detector Photons at Optimal: {int(results['detector_photons'][optimal_idx])}")
    print(f"Forward Photons at Optimal: {int(results['forward_photons'][optimal_idx])}")
    
    print("\n--- Top 3 Configurations ---")
    sorted_indices = np.argsort(results['efficiency'])[::-1][:3]
    for i, idx in enumerate(sorted_indices, 1):
        print(f"{i}. Gap={results['gap'][idx]:.0f} cm, "
              f"Efficiency={results['efficiency'][idx]:.2f}%, "
              f"Photons={int(results['detector_photons'][idx])}")
    
    print("\n--- Gap Range Analysis ---")
    print(f"Minimum Efficiency: {np.min(results['efficiency']):.2f}% "
          f"at {results['gap'][np.argmin(results['efficiency'])]:.0f} cm")
    print(f"Maximum Efficiency: {np.max(results['efficiency']):.2f}% "
          f"at {results['gap'][np.argmax(results['efficiency'])]:.0f} cm")
    print(f"Average Efficiency: {np.mean(results['efficiency']):.2f}%")
    
    print("\n" + "="*60)

if __name__ == "__main__":
    if len(sys.argv) > 1:
        filename = sys.argv[1]
    else:
        filename = 'gap_scan_results.txt'
    
    try:
        results = read_gap_scan_results(filename)
        
        output_prefix = filename.replace('.txt', '')
        optimal_gap, optimal_eff = plot_gap_scan_results(results, output_prefix)
        
        print_recommendations(results)
        
        plt.show()
        
    except FileNotFoundError:
        print(f"Error: {filename} not found!")
        print("Please run the gap scan simulation first.")
    except Exception as e:
        print(f"Error: {e}")

File: analyze_gap_scan.cpp
/*
 * DAMSA Gap Scan Results Analysis
 * Reads gap scan results and creates ROOT visualization plots.
 *
 * Usage:
 *   ./build/analyze_gap_scan [gap_scan_results.txt]
 */

#include "TCanvas.h"
#include "TGraph.h"
#include "TMultiGraph.h"
#include "TAxis.h"
#include "TLegend.h"
#include "TLatex.h"
#include "TLine.h"
#include "TStyle.h"
#include "TPaveText.h"
#include "TROOT.h"

#include <fstream>
#include <sstream>
#include <string>
#include <vector>
#include <iostream>
#include <algorithm>
#include <numeric>

struct GapScanData {
    std::vector<double> gap;
    std::vector<double> csi_z;
    std::vector<double> target_photons;
    std::vector<double> target_neutrons;
    std::vector<double> detector_photons;
    std::vector<double> detector_neutrons;
    std::vector<double> forward_photons;
    std::vector<double> efficiency;
};

GapScanData ReadGapScanResults(const std::string& filename) {
    GapScanData data;
    std::ifstream file(filename);
    if (!file.is_open()) {
        std::cerr << "Error: " << filename << " not found!\n";
        std::cerr << "Please run the gap scan simulation first.\n";
        std::exit(1);
    }

    std::string line;
    while (std::getline(file, line)) {
        // Trim whitespace
        size_t start = line.find_first_not_of(" \t\r\n");
        if (start == std::string::npos) continue;
        line = line.substr(start);

        // Skip comments
        if (line[0] == '#') continue;

        std::istringstream iss(line);
        double v[8];
        int count = 0;
        while (count < 8 && (iss >> v[count])) count++;

        if (count >= 8) {
            data.gap.push_back(v[0]);
            data.csi_z.push_back(v[1]);
            data.target_photons.push_back(v[2]);
            data.target_neutrons.push_back(v[3]);
            data.detector_photons.push_back(v[4]);
            data.detector_neutrons.push_back(v[5]);
            data.forward_photons.push_back(v[6]);
            data.efficiency.push_back(v[7]);
        }
    }

    if (data.gap.empty()) {
        std::cerr << "Error: No data found in file!\n";
        std::exit(1);
    }
    return data;
}

// Helper to style a TGraph
void StyleGraph(TGraph* g, Color_t color, int marker, const char* xtitle, const char* ytitle) {
    g->SetLineColor(color);
    g->SetLineWidth(2);
    g->SetMarkerColor(color);
    g->SetMarkerStyle(marker);
    g->SetMarkerSize(1.2);
    g->GetXaxis()->SetTitle(xtitle);
    g->GetYaxis()->SetTitle(ytitle);
    g->GetXaxis()->SetTitleFont(42);
    g->GetYaxis()->SetTitleFont(42);
    g->GetXaxis()->SetLabelFont(42);
    g->GetYaxis()->SetLabelFont(42);
    g->GetXaxis()->SetTitleSize(0.05);
    g->GetYaxis()->SetTitleSize(0.05);
}

void PlotGapScanResults(const GapScanData& data, const std::string& outputPrefix) {
    gStyle->SetOptStat(0);
    gStyle->SetPadTickX(1);
    gStyle->SetPadTickY(1);
    gStyle->SetCanvasDefW(1600);
    gStyle->SetCanvasDefH(1000);

    int N = data.gap.size();

    // Find optimal point
    int optIdx = std::distance(data.efficiency.begin(),
                               std::max_element(data.efficiency.begin(), data.efficiency.end()));
    double optGap = data.gap[optIdx];
    double optEff = data.efficiency[optIdx];

    TCanvas* c = new TCanvas("c_gap", "DAMSA Gap Scan Results", 1600, 1000);
    c->Divide(3, 2, 0.01, 0.02);

    // --- Pad 1: Efficiency vs Gap ---
    c->cd(1);
    gPad->SetLeftMargin(0.14);
    gPad->SetRightMargin(0.05);
    gPad->SetGrid();
    TGraph* g1 = new TGraph(N, data.gap.data(), data.efficiency.data());
    StyleGraph(g1, kBlue, 20, "Gap Distance [cm]", "Efficiency [%]");
    g1->SetTitle("Photon Detection Efficiency vs Gap");
    g1->Draw("ALP");
    TLine* vline1 = new TLine(optGap, g1->GetYaxis()->GetXmin(), optGap, optEff);
    vline1->SetLineColor(kRed);
    vline1->SetLineStyle(2);
    vline1->Draw();
    // Optimal star marker
    TGraph* gStar = new TGraph(1);
    gStar->SetPoint(0, optGap, optEff);
    gStar->SetMarkerStyle(29);
    gStar->SetMarkerSize(3.0);
    gStar->SetMarkerColor(kRed);
    gStar->Draw("P SAME");
    TLegend* leg1 = new TLegend(0.50, 0.75, 0.93, 0.90);
    leg1->SetBorderSize(0);
    leg1->SetFillStyle(0);
    leg1->SetTextFont(42);
    leg1->AddEntry(g1, "Photon Efficiency", "lp");
    leg1->AddEntry(gStar, Form("Optimal: %.0f cm", optGap), "p");
    leg1->Draw();

    // --- Pad 2: Detector Photons vs Gap ---
    c->cd(2);
    gPad->SetLeftMargin(0.14);
    gPad->SetRightMargin(0.05);
    gPad->SetGrid();
    TGraph* g2 = new TGraph(N, data.gap.data(), data.detector_photons.data());
    StyleGraph(g2, kGreen + 2, 20, "Gap Distance [cm]", "Number of Photons");
    g2->SetTitle("Photons Reaching Detector vs Gap");
    g2->Draw("ALP");
    TLine* vline2 = new TLine(optGap, g2->GetYaxis()->GetXmin(), optGap, data.detector_photons[optIdx]);
    vline2->SetLineColor(kRed);
    vline2->SetLineStyle(2);
    vline2->Draw();

    // --- Pad 3: Forward Photons vs Gap ---
    c->cd(3);
    gPad->SetLeftMargin(0.14);
    gPad->SetRightMargin(0.05);
    gPad->SetGrid();
    TGraph* g3 = new TGraph(N, data.gap.data(), data.forward_photons.data());
    StyleGraph(g3, kViolet, 20, "Gap Distance [cm]", "Number of Forward Photons");
    g3->SetTitle("Forward Photons (0-20#circ) vs Gap");
    g3->Draw("ALP");
    TLine* vline3 = new TLine(optGap, g3->GetYaxis()->GetXmin(), optGap, data.forward_photons[optIdx]);
    vline3->SetLineColor(kRed);
    vline3->SetLineStyle(2);
    vline3->Draw();

    // --- Pad 4: Detector Neutrons vs Gap ---
    c->cd(4);
    gPad->SetLeftMargin(0.14);
    gPad->SetRightMargin(0.05);
    gPad->SetGrid();
    TGraph* g4 = new TGraph(N, data.gap.data(), data.detector_neutrons.data());
    StyleGraph(g4, kOrange + 1, 20, "Gap Distance [cm]", "Number of Neutrons");
    g4->SetTitle("Neutrons Reaching Detector vs Gap");
    g4->Draw("ALP");
    TLine* vline4 = new TLine(optGap, g4->GetYaxis()->GetXmin(), optGap, data.detector_neutrons[optIdx]);
    vline4->SetLineColor(kRed);
    vline4->SetLineStyle(2);
    vline4->Draw();

    // --- Pad 5: Target Exit vs Detector Photons ---
    c->cd(5);
    gPad->SetLeftMargin(0.14);
    gPad->SetRightMargin(0.05);
    gPad->SetGrid();
    TGraph* g5a = new TGraph(N, data.gap.data(), data.target_photons.data());
    TGraph* g5b = new TGraph(N, data.gap.data(), data.detector_photons.data());
    StyleGraph(g5a, kBlue, 20, "Gap Distance [cm]", "Number of Photons");
    g5b->SetLineColor(kGreen + 2);
    g5b->SetLineWidth(2);
    g5b->SetMarkerColor(kGreen + 2);
    g5b->SetMarkerStyle(22);
    g5b->SetMarkerSize(1.2);

    TMultiGraph* mg = new TMultiGraph();
    mg->Add(g5a, "LP");
    mg->Add(g5b, "LP");
    mg->SetTitle("Target Exit vs Detector Photons;Gap Distance [cm];Number of Photons");
    mg->Draw("A");
    mg->GetXaxis()->SetTitleFont(42);
    mg->GetYaxis()->SetTitleFont(42);
    mg->GetXaxis()->SetTitleSize(0.05);
    mg->GetYaxis()->SetTitleSize(0.05);
    TLine* vline5 = new TLine(optGap, mg->GetYaxis()->GetXmin(), optGap, data.target_photons[optIdx]);
    vline5->SetLineColor(kRed);
    vline5->SetLineStyle(2);
    vline5->Draw();
    TLegend* leg5 = new TLegend(0.50, 0.75, 0.93, 0.90);
    leg5->SetBorderSize(0);
    leg5->SetFillStyle(0);
    leg5->SetTextFont(42);
    leg5->AddEntry(g5a, "Target Exit Photons", "lp");
    leg5->AddEntry(g5b, "Detector Photons", "lp");
    leg5->Draw();

    // --- Pad 6: Optimal configuration summary ---
    c->cd(6);
    gPad->SetLeftMargin(0.05);
    gPad->SetRightMargin(0.05);

    TPaveText* pt = new TPaveText(0.05, 0.05, 0.95, 0.95, "NDC");
    pt->SetBorderSize(1);
    pt->SetFillColor(kWhite);
    pt->SetTextFont(42);
    pt->SetTextAlign(12);
    pt->SetTextSize(0.06);

    pt->AddText("#bf{Optimal Configuration}");
    pt->AddText("");
    pt->AddText(Form("Optimal Gap:              %.0f cm", optGap));
    pt->AddText(Form("Optimal Efficiency:     %.2f%%", optEff));
    pt->AddText(Form("CsI Position:              %.1f cm", data.csi_z[optIdx]));
    pt->AddText(Form("Detector Photons:      %.0f", data.detector_photons[optIdx]));
    pt->AddText(Form("Forward Photons:        %.0f", data.forward_photons[optIdx]));
    pt->Draw();

    // Title
    TLatex title;
    title.SetNDC();
    title.SetTextFont(42);
    title.SetTextSize(0.04);

    std::string outfile = outputPrefix + "_analysis.png";
    c->SaveAs(outfile.c_str());
    std::cout << "Analysis plot saved as: " << outfile << "\n";

    delete c;
}

void PrintRecommendations(const GapScanData& data) {
    int N = data.gap.size();

    int optIdx = std::distance(data.efficiency.begin(),
                               std::max_element(data.efficiency.begin(), data.efficiency.end()));

    std::cout << "\n" << std::string(60, '=') << "\n";
    std::cout << "DAMSA GAP SCAN RECOMMENDATIONS\n";
    std::cout << std::string(60, '=') << "\n";
    std::cout << "\nOptimal Gap Distance: " << data.gap[optIdx] << " cm\n";
    std::cout << "Corresponding CsI Center Z: " << data.csi_z[optIdx] << " cm\n";
    std::cout << "Maximum Efficiency: " << data.efficiency[optIdx] << "%\n";
    std::cout << "Detector Photons at Optimal: " << (int)data.detector_photons[optIdx] << "\n";
    std::cout << "Forward Photons at Optimal: " << (int)data.forward_photons[optIdx] << "\n";

    // Top 3
    std::vector<int> indices(N);
    std::iota(indices.begin(), indices.end(), 0);
    std::sort(indices.begin(), indices.end(),
              [&](int a, int b) { return data.efficiency[a] > data.efficiency[b]; });

    std::cout << "\n--- Top 3 Configurations ---\n";
    for (int i = 0; i < std::min(3, N); ++i) {
        int idx = indices[i];
        std::cout << (i + 1) << ". Gap=" << data.gap[idx] << " cm, "
                  << "Efficiency=" << data.efficiency[idx] << "%, "
                  << "Photons=" << (int)data.detector_photons[idx] << "\n";
    }

    double minEff = *std::min_element(data.efficiency.begin(), data.efficiency.end());
    double maxEff = *std::max_element(data.efficiency.begin(), data.efficiency.end());
    double avgEff = std::accumulate(data.efficiency.begin(), data.efficiency.end(), 0.0) / N;
    int minIdx = std::distance(data.efficiency.begin(),
                               std::min_element(data.efficiency.begin(), data.efficiency.end()));

    std::cout << "\n--- Gap Range Analysis ---\n";
    std::cout << "Minimum Efficiency: " << minEff << "% at " << data.gap[minIdx] << " cm\n";
    std::cout << "Maximum Efficiency: " << maxEff << "% at " << data.gap[optIdx] << " cm\n";
    std::cout << "Average Efficiency: " << avgEff << "%\n";

    std::cout << "\n" << std::string(60, '=') << "\n";
}

int main(int argc, char** argv) {
    gROOT->SetBatch(true);

    std::string filename = "gap_scan_results.txt";
    if (argc > 1) filename = argv[1];

    GapScanData data = ReadGapScanResults(filename);

    // Output prefix: strip .txt extension
    std::string prefix = filename;
    size_t dotPos = prefix.rfind(".txt");
    if (dotPos != std::string::npos) prefix = prefix.substr(0, dotPos);

    PlotGapScanResults(data, prefix);
    PrintRecommendations(data);

    return 0;
}
